### DISCUSSION TITLE
Alpha MwV2 vs MwV3

### DISCUSSION THREAD

**Discussion by seogrady** [oscartbeaumont/rspc#185] [rspc Help] [answered]
Hey Oscar! I've been playing around with Rspc and I really like it. With the alpha version, how can I return a Result<> from a middleware function without having to use MwV3? I don't need to capture extra arguments in the middleware and can authenticate using headers/cookie.  Are you able to provide an example?

**Comment by oscartbeaumont** [comment, marked as answer]
The difference between `MwV2` & `MwV3` isn't actually the arguments, it's just how the arguments are _constrained_. They both need the two arguments and the same function signature to be valid (The advantage of the unconstrained type is that you can provide non-closure types but this is mostly for super advanced stuff). The reason `MwV2` exists comes down to [this](https://users.rust-lang.org/t/rust-not-able-to-infer-closure-argument-type-when-implmenting-trait-on-closure/76942) ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f69839e43e00050282f2137721b043d7)) and I am still working on a coming up with a solution to this because it's really not great DX but Rust makes it very difficult.

Would be good for you to send through the code you've got cause I don't know if you're running on the latest crates.io release, main, or the v1-alpha. All of them handle stuff pretty differently. Can't wait till v1 is properly released but I just don't have the time right now to get it over the line.

**Reply by seogrady** [reply]
Thanks for the quick reply @oscartbeaumont ! Yeah it's a bit confusing with `MwV2` & `MwV3`. I was having a look at the spacedrive usage of Rspc and saw they created a `MwArgMapperMiddleware::<LibraryArgsLike>::new()` that returns a `Result<>`. As there wasn't an example that returned a `Result<>`, I thought that `MwV3` was required.

FYI – I checked out the repo and played around with the tests you have in `src/alpha/mod.rs`. Wrapping the result in `Ok()` compiles so 👍 . I think I just got a little lost looking through the code.

```rust
pub fn library<TLCtx>() -> impl MwV2<TLCtx, NewCtx = (TLCtx, i32)>
where
    TLCtx: Send + Sync + 'static,
{
    |mw, ctx| async move { Ok(mw.next((ctx, 42))) }
}
```

Cheers mate!

**Reply by oscartbeaumont** [reply]
I'm closing this as v0.4.0 comes with a new middleware syntax which shouldn't have this issue.
